### DISCUSSION TITLE
Migration Guide?

### DISCUSSION THREAD

**Discussion by Duncanduncability** [rabbitmq/rabbitmq-dotnet-client#1720] [Q&A] [answered]
Hi, I'm looking for advice for how to migrate from v6.8.1 to the new v7. Any idea when new docs and examples will be available?

**Comment by lukebakken** [comment]
Hello, thanks for your interest.

As of now, the best examples are the demo applications and test suites. I'm in the process of updating the [tutorial code](https://github.com/rabbitmq/rabbitmq-tutorials/pull/473) as well as the [tutorials on the website](https://github.com/rabbitmq/rabbitmq-website/pull/2049).

In short, though, these are the main changes:
* Rename `IModel` to `IChannel` everywhere.
* Correctly use `async` / `await` with the updated API.
* If you are using publisher confirmations, the library now efficiently tracks them for you. See [this code](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/Applications/PublisherConfirms/PublisherConfirms.cs#L110-L171) for an example.
* Create `BasicProperties` using `new` instead of the old way.

**Reply by michaelklishin** [reply]
Yay, tutorial 7 for .NET is coming :D

**Reply by mgroetan** [reply]
I've found that the "EventingBasicConsumer" class has been renamed to the not so logically "**Async**EventingBasicConsumer".

**Reply by lukebakken** [reply]
Yes, we followed TPL guidelines to ensure that methods and other entities use the `Async` suffix or (in this case), prefix.

**Reply by affansamad93** [reply]
In case of any error, I need to apply retry logic. How do I achieve it in RabbitMQ Client 7+ version?

**Reply by lukebakken** [reply]
@affansamad93 if you'd like assistance, the best way is to provide code that I can clone to see what you have tried.

**Comment by lukebakken** [comment]
@Duncanduncability - I just finished updating and merging the tutorial code - https://github.com/rabbitmq/rabbitmq-tutorials/tree/main/dotnet

**Comment by lechu445** [comment]
The other change worth mentioning in the migration guide is when you used `IModel.ConfirmSelect()` now you should set `CreateChannelOptions.PublisherConfirmationsEnabled` to `true` in `IConnection.CreateChannelAsync`. 

This change was introduced in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1687.

**Comment by eithe** [comment]
I am trying to update our tests that uses Moq, Xunit and FluentAssertions, but QueueDeclareAsync is now an IChannel extension method which makes it more difficult to mock. Was this by intention? 

<details>

```
public class RabbitMqClientTest
{
    private (Mock<IRabbitMqConnection>, IMessageQueue) GetMocks()
    {
        var mockRabbitMqConnection = new Mock<IRabbitMqConnection>();
        var mockChannel = new Mock<IChannel>();
        var mockProperties = new Mock<IBasicProperties>();
        mockChannel.Setup(x => x.IsOpen).Returns(true);
        mockChannel.Setup(x => x.GetNextPublishSequenceNumberAsync(It.IsAny<CancellationToken>())).ReturnsAsync(1UL);
        mockChannel.Setup(x => x.QueueDeclareAsync(It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<bool>(), It.IsAny<bool>(), It.IsAny<IDictionary<string, object>>(), false, It.IsAny<CancellationToken>()))
            .ReturnsAsync(new QueueDeclareOk("test_queue", 1, 1));
        mockChannel.Setup(x => x.BasicAckAsync(It.IsAny<ulong>(), It.IsAny<bool>(), It.IsAny<CancellationToken>())).Returns(new ValueTask());
        mockChannel.SetupAdd(x => x.BasicAcksAsync += It.IsAny<AsyncEventHandler<BasicAckEventArgs>>());
        mockChannel.SetupAdd(x => x.BasicNacksAsync += It.IsAny<AsyncEventHandler<BasicNackEventArgs>>());
        mockChannel.Setup(x => x.BasicPublishAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<BasicProperties>(), It.IsAny<ReadOnlyMemory<byte>>(), It.IsAny<CancellationToken>())).Returns(new ValueTask());
        mockChannel.Setup(x => x.BasicConsumeAsync(It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<bool>(),
            It.IsAny<IDictionary<string, object>>(), It.IsAny<IAsyncBasicConsumer>(), It.IsAny<CancellationToken>())).ReturnsAsync("test_consumer");
        mockRabbitMqConnection.Setup(x => x.Channel).Returns(mockChannel.Object);
        var sut = new RabbitMqClient(mockRabbitMqConnection.Object, Mock.Of<ILogger<RabbitMqClient>>(), Mock.Of<IRedisClient>());
        return (mockRabbitMqConnection, sut);
    }

    private IMessageQueue GetMessageQueue()
    {
        var mocks = GetMocks();
        IMessageQueue sut = mocks.Item2;
        return sut;
    }

    [Fact]
    public void CanInstantiateClient()
    {
        var mocks = GetMocks();
        IRabbitMqConnection rmqc = mocks.Item1.Object;
        rmqc.Should().NotBeNull();
        IMessageQueue sut = mocks.Item2;
        sut.Should().NotBeNull();
        sut.Should().BeOfType<RabbitMqClient>();
    }

    [Fact]
    public async Task CanProduce()
    {
        var sut = GetMessageQueue();
        var result = await sut.Produce(MessageConstants.TestExchange, new Message<string> { Caller = "test", Content = "Test string", Id = Guid.NewGuid() }, false);

        result.Should().BeOfType<MessageQueueResult>();
        result.IsError.Should().BeFalse();
        result.SequenceNumber.Should().Be(1);
    }

    [Fact]
    public async void CanSetupConsumer()
    {
        var sut = GetMessageQueue();
        var result = await sut.Consume<string>(MessageConstants.TestExchange, MessageConstants.TestQueue, DoWork);
        result.Should().NotBeNull();
        result.Should().Be("test_consumer");
    }

    private async Task<bool> DoWork<T>(IMessage<T> message, CancellationToken cancellationToken)
    {
        message.Should().NotBeNull();
        return true;
    }
}
```
</details>

Which results in 

```
System.NotSupportedException : Unsupported expression: x => x.QueueDeclareAsync(It.IsAny<string>(), It.IsAny<bool>(), It.IsAny<bool>(), It.IsAny<bool>(), It.IsAny<IDictionary<string, object>>(), False, It.IsAny<CancellationToken>())
Extension methods (here: IChannelExtensions.QueueDeclareAsync) may not be used in setup / verification expressions.
```

**Reply by lukebakken** [reply]
Use `IChannel.QueueDeclareAsync` - https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/v7.0.0/projects/RabbitMQ.Client/IChannel.cs#L341-L354

**Comment by lechu445** [comment]
> I'm looking for advice for how to migrate from v6.8.1 to the new v7

@Duncanduncability the migration guide you can find here https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/v7-MIGRATION.md

**Reply by lukebakken** [reply]
Thanks, and I will update that guide this week with what other users have mentioned in this discussion.

**Reply by michaelklishin** [reply]
And everyone is welcome to contribute to it :)

**Reply by lechu445** [reply]
Some changes that I found during migration:
- all IBasicConsumer should be replaced with IAsyncBasicConsumer (IMO this breaking change was unnecessary)
- all IModel should be replaced with IChannel
- IChannel does not contain `BasicRecoverOk` event that was existing in IModel
- for AsyncEventingBasicConsumer all events now have -Async suffix (example `ShutdownAsync`)
- AsyncEventingBasicConsumer has removed `ConsumerCancelled` event
- BasicDeliverEventArgs changed from property initialization to constructor initialization
- `IChannel.BasicPublishAsync` takes a new parameter `mandatory`

**Reply by michaelklishin** [reply]
`BasicRecoverOk` can be added back but `basic.recover` has had no real purpose in the protocol for over a decade. You arguably should stop using it instead.

**Reply by lukebakken** [reply]
> basic.recover has had no real purpose in the protocol for over a decade

That's why I removed `BasicRecoverOk`. I'll make a note about it. Thank you for following up @lechu445. How has version 7 been performing for you?

**Reply by lechu445** [reply]
I have a low throughput app therefore I din't observe any significant performance gain. I had no issues so far. I am waiting for release HealthChecks.RabbitMQ with the upgraded version https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/2323.

**Comment by lechu445** [comment]
It is also not clear how to replace IModel.WaitForConfirmsOrDie with provided timeout. Should it be set to IChannel.ContinuationTimeout or is it a different thing?

I found here https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1234#issuecomment-1192771092 only encougare to replace IModel.WaitForConfirmsOrDie with publisher confirms  but its documentation does not mention about the timeout.

**Reply by lukebakken** [reply]
In the 6.x library, `IModel.WaitForConfirmsOrDie` _depends_ on publisher confirmations - you can't call `IModel.WaitForConfirmsOrDie` without first having called `IModel.ConfirmSelect`.

In version 7, you enable publisher confirmations when you create the channel, via `CreateChannelOptions`, then, when you call `BasicPublishAsync`, pass a `CancellationToken` with your desired timeout value for receiving the single (or multiple) confirmations.

Please see [this example code](https://github.com/lukebakken/rabbitmq-dotnet-client-1721/blob/main/Program.cs) that I put together to address [this comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1721#discussioncomment-11206253).

I will make a note to clarify how to set a confirmation timeout in version 7 of this library.

**Comment by niemyjski** [comment]
This change is kind of confusing looking at the 7.0 breaking changes.

```cs
_factory = new ConnectionFactory
        {
            Uri = new Uri(options.ConnectionString),
            AutomaticRecoveryEnabled = true,
            DispatchConsumersAsync = true
        };
```

DispatchConsumersAsync where did this get moved to now?

**Reply by lukebakken** [reply]
`DispatchConsumersAsync` is no longer necessary, as all consumers are async in version 7. Thanks for pointing this out.

**Comment by ZajacPiotr98** [comment]
Hello.

Is this appropriate to pass cancellation token to the IChannel `BasicConsumeAsync` or any other method that invokes `ModelSendAsync`? When Cancellation request and task cancelled exception is thrown while Connection class is processing an action in the MainLoop, then connection is closing 
`CloseReason: AMQP close-reason, initiated by Library, code=541, text='A task was canceled.', classId=0, methodId=0, exception=System.Threading.Tasks.TaskCanceledException: A task was canceled.    at RabbitMQ.Client.Impl.BasicConsumeAsyncRpcContinuation.HandleCommandAsync(IncomingCommand cmd)    at RabbitMQ.Client.Impl.Channel.HandleCommandAsync(IncomingCommand cmd, CancellationToken cancellationToken)    at RabbitMQ.Client.Framing.Connection.ProcessFrameAsync(InboundFrame frame, CancellationToken cancellationToken)    at RabbitMQ.Client.Framing.Connection.ReceiveLoopAsync(CancellationToken mainLoopCancellationToken)    at RabbitMQ.Client.Framing.Connection.MainLoop()`

**Reply by lukebakken** [reply]
> Is this appropriate to pass cancellation token to the IChannel BasicConsumeAsync or any other method that invokes ModelSendAsync? 

Yes you should be able to do this without closing the connection. Thanks for reporting the issue.

**Reply by lukebakken** [reply]
https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1750

**Reply by ZajacPiotr98** [reply]
I think passing already cancelled cancellation token will trigger exception already on the `await _rpcSemaphore.WaitAsync(k.CancellationToken)`. The problem occurs when it's passed to the Connection

**Reply by lukebakken** [reply]
@ZajacPiotr98 would you mind adding that comment to the issue itself, and continuing all discussion there? Thank you.

https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1750

**Comment by gokcengintertech** [comment]
Maybe it's just me but I haven't had a clear understanding of how to replace those structures in 7.0.0?

- WaitForConfirms / ConfirmSelect
- ConsumerCancelled

Especially when it comes to WaitForConfirms, I was able to configure the client while sending a message in 6.x and now it seems it's moved to CreateChannelOptions so I kind of feel it can only be configured per channel not the message. Please correct me if I'm wrong there.

**Reply by lukebakken** [reply]
> it can only be configured per channel not the message

It has always been per-channel. In 6.x you could only call `WaitForConfirms` if `ConfirmSelect` had been called on the channel first.

When the channel has confirmations enabled, and you `await` the result of `BasicPublishAsync` you are awaiting the confirmation.

See [this comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1720#discussioncomment-11255938) and linked code.

As for `ConsumerCancelled`, it appears we re-named that event:

https://github.com/search?q=repo%3Arabbitmq%2Frabbitmq-dotnet-client%20ConsumerCancelled&type=code

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/Test/Integration/TestAsyncConsumerCancelNotify.cs#L77-L86

**Comment by jehrenzweig-captionhealth** [comment]
I'm trying to figure out how to upgrade from 6.8.1 to 7.0.0, & I could use some pointers about how to replace references to the following:

* `IBasicPublishBatch` interface
* `ConnectionFactory.DispatchConsumersAsync` property
* `IModel.CreateBasicPublishBatch()` method

**Reply by lukebakken** [reply, marked as answer]
* https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1721#discussioncomment-11206253
* https://github.com/rabbitmq/rabbitmq-dotnet-client/discussions/1720#discussioncomment-11373986

**Reply by jehrenzweig-captionhealth** [reply]
Thank you!
